### DISCUSSION TITLE
Debug Write From Source

### DISCUSSION THREAD

**Discussion by nick-sasquatch** [JuulLabs/kable#611] [Q&A] [answered]
Hi all--I'm running into an occasional issue where a `peripheral.write()` on Android doesn't return, and I'd like to try to debug it.

This is a snippet of the usage:

```kotlin
    fun writeCharacteristic(data: ByteArray) {
        if(isConnected.get()) {
            owner.lifecycleScope.launch(Dispatchers.IO) {
                try {
                    Log.i(LOG_TAG, "$peripheral about to write")
                    peripheral.write(CUSTOM_RX_CHARACTERISTIC, data, WriteType.WithoutResponse)
                    Log.i(LOG_TAG, "$peripheral write complete")
                } catch(e: Exception)  {
                    Log.w(LOG_TAG, "Unable to write to characteristic.")
                }
            }
        }
    }
```

I haven't characterized the problem completely yet, but I've chased it down to the "about to write" message no longer being followed by a "write complete" message--it appears that the write is hung somehow. Once that's happened once, it'll happen on subsequent writes.

I'm trying to debug this myself for a bit in an effort to understand if it's our business logic somehow putting us in this spot, but I'm having a heck of a time trying to bring in kable as an external module to debug from source. I've tried "File->New->Import module..." in Android studio and variants of the following in our gradle files with no avail:

```
// settings.gradle
include(":com.juul.kable")
project(":com.juul.kable").projectDir = file("./kable")
```

```
// build.gradle
implementation(project(":com.juul.kable"))
```

Do you have any hints on how to do that? Or ideas about what might be happening to the write? Thanks a ton and have a great holiday.

**Comment by twyatt** [comment, marked as answer]
I believe the easiest way to test locally is via Maven local install as described in https://github.com/JuulLabs/kable/pull/494#issuecomment-1613724319.

**Reply by nick-sasquatch** [reply]
Sounds great. That worked like a charm, thank you.

**Comment by twyatt** [comment]
How large is the `ByteArray` (`data`) that you're writing? If it exceeds the MTU, you may be hitting #298?

**Reply by nick-sasquatch** [reply]
Thanks again for the quick reply on this--that write's much smaller than the MTU.

**Reply by twyatt** [reply]
Hmm. Well keep me posted then on what you find in your debugging. Good luck!

**Reply by miwright2** [reply]
Hey @twyatt -- it looks to me when reading the remote RSSI that Android sometimes fails to callback to Kable.  When that happens, Kable deadlocks, waiting for that callback/response to come back, and then subsequent peripheral writes and RSSI reads fail.  This is on a Pixel 7 current Android.  Let me know if you want me to provide some specific pointers into the code to better explain what I'm seeing.

**Reply by twyatt** [reply]
Oof. Thanks for finding that issue. Can you [create an issue](https://github.com/JuulLabs/kable/issues/new), please? And I'll try to find some time over the winter break to look into it.

**Reply by miwright2** [reply]
Sure thing -- I think our use case is a little weird.  We have some features that depend on proximity while connected, so need to poll on the RSSI value.
